Add `reflog` support in `command` menu

pmiossec opened the pull request:
To show advancement and gather feedback....

pmiossec commented:
Seems a good start for me.
If someone wants to review it....

pmiossec commented:
![reflog_menu](https://cloud.githubusercontent.com/assets/460196/17537656/d5688a7a-5e9e-11e6-97d9-e3a999b8e026.png)
![reflog_commit_list](https://cloud.githubusercontent.com/assets/460196/17537658/d81ee67e-5e9e-11e6-8858-c4e89f35d1be.png)

PS: I don't tthink that's my fault if the AppVeyor build is broken... I will be pleased if someone could run it again ;)

pmiossec commented:
AppVeyor build is passing....

pmiossec commented:
If someone could do a review or test because the development is finished in my opinion and could be merged without introducing too much problem (there is no big changes here...).

It's for me a good git feature to know when you are a git beginner and it could save you more than once. I think it make sense to have it inside GitExtensions

jbialobr commented on GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs:
```diff
@@ -0,0 +1,173 @@
+{
+    public partial class FormReflog : GitModuleForm
+    {
+        private readonly TranslationString _continueResetCurrentBranchText = new TranslationString("Are you sure to want to reset the current branch to this commit?");
```
Are you sure you want to reset the current branch to this commit?

jbialobr commented on GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs:
```diff
@@ -0,0 +1,173 @@
+    public partial class FormReflog : GitModuleForm
+    {
+        private readonly TranslationString _continueResetCurrentBranchText = new TranslationString("Are you sure to want to reset the current branch to this commit?");
+        private readonly TranslationString _continueResetCurrentBranchCaptionText = new TranslationString("Reset current branch to a specific commit");
```
Reset the current branch to a specific commit.

jbialobr commented on GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs:
```diff
@@ -0,0 +1,173 @@
+        {
+            _currentBranch = UICommands.Module.GetSelectedBranch();
+            linkCurrentBranch.Text = "current branch (" + _currentBranch + ")";
+            _isBranchCheckedOut = _currentBranch != "(no branch)";
```
GitModule.IsDetachedHead

pmiossec replied:
I had to add this method to the `IGitModule` to be able to use it...

jbialobr commented on GitUI/CommandsDialogs/RefLogDialog/BranchNameForm.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace GitUI.CommandsDialogs.RefLogDialog
+{
+    public partial class BranchNameForm : Form
```
Branch name should be normalized according to the settings. See the IGitBranchNameNormaliser and its usages for details.

jbialobr replied:
Is it possible to use the existsing Create branch dialog here?

pmiossec replied:
>Branch name should be normalized according to the settings.

Done.

>Is it possible to use the existsing Create branch dialog here?

I'm not really convinced by that idea. This dialog is too complex for just creating a branch in this specific case (I myself not very confident on the result if you begin to change some options in the form before creating the branche, in the context of using the reflog).

I have made the development in another branch ( https://github.com/pmiossec/gitextensions/tree/reflog_reuse_create_branch_form ) if you want to test. Let me know...

jbialobr replied:
Thanks! I tested it and I find it a better solution. The existing Create Branch dialog is well known to users so they should have no problem with the additional complexity. I would only uncheck the Checkout After Create checkbox when using from the RefLogDialog.

pmiossec commented:
>Only reviewed, not tested yet.

That's what I ask for. Thanks!

I have made all the fixes...

jbialobr commented:
The regex does not match this line:
`3750ade pr/3242@{0}: fetch origin pull/3242/head:pr/3242: storing ref`

pmiossec commented:
>The regex does not match this line:

I hope that's fixed now...

jbialobr commented on GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs:
```diff
@@ -0,0 +1,173 @@
+            var branches = new List<string> { "HEAD" };
+            branches.AddRange(UICommands.Module.GetRefs(false, true).Select(r => r.Name));
+            Branches.DataSource = branches;
+            DisplayRefLog();
```
It seems that assigning the DataSource property, in the line above, triggers the Branches_SelectedIndexChanged. That causes the reflog command being executed twice at creating the dialog.

pmiossec commented:
@jbialobr I have made the fixes in this PR

jbialobr commented:
Thank you! It is a very useful tool, I have already used it a few times.

pmiossec commented:
>Thank you! It is a very useful tool, I have already used it a few times.

Especially useful for beginners when they mess their repository history. I think that's one of the first notion that a beginner should learn when learning git because that permit to do errors but recover from it and retry...

PS: I have squashed the history to make it cleaner (and cleaned a little the code using https://github.com/dotnet/codeformatter `CodeFormatter.exe /rule-:FieldNames /nounicode /nocopyright .\GitExtensions.VS2015.sln` ).

Ready for merging... ;)

l-inc commented:
A nice feature, but has a couple of problems:
1) In mono `gridReflog.Rows[hit.RowIndex].Selected = true;` does not deselect a previously selected row. I used to apply a little patch to workaround that:
```
diff --git a/GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs b/GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs
index f917ab9..f57243d 100644
--- a/GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs
+++ b/GitUI/CommandsDialogs/RefLogDialog/FormReflog.cs
@@ -16,6 +16,7 @@ public partial class FormReflog : GitModuleForm
 
         private string _currentBranch;
         private bool _isBranchCheckedOut;
+        private int _lastHitRowIndex;
 
         public FormReflog(GitUICommands uiCommands)
             : base(uiCommands)
@@ -29,6 +30,7 @@ private void FormReflog_Load(object sender, EventArgs e)
             _currentBranch = UICommands.Module.GetSelectedBranch();
             linkCurrentBranch.Text = "current branch (" + _currentBranch + ")";
             _isBranchCheckedOut = _currentBranch != "(no branch)";
+            _lastHitRowIndex = 0;
             resetCurrentBranchOnThisCommitToolStripMenuItem.Enabled = _isBranchCheckedOut;
 
             var branches = new List<string> { "HEAD" };
@@ -146,7 +148,9 @@ private void gridReflog_MouseMove(object sender, MouseEventArgs e)
             DataGridView.HitTestInfo hit = gridReflog.HitTest(e.X, e.Y);
             if (hit.Type == DataGridViewHitTestType.Cell)
             {
-                gridReflog.Rows[hit.RowIndex].Selected = true;
+                gridReflog.Rows[_lastHitRowIndex].Selected = false;
+                _lastHitRowIndex = hit.RowIndex;
+                gridReflog.Rows[_lastHitRowIndex].Selected = true;
             }
         }
     }
```
2) It's a bit dangerous. I lost uncommited changes a couple of times when resetting to a previous state. The standard GitExtensions interface for resetting a branch provides Soft, Mixed and Hard options (defaulting to Mixed) to prevent such unintended data losses.

pmiossec commented:
@l-inc Thanks for the head up!

I have updated the PR.

1) I have included your fix (even if I don't understand everything...)
2) I have added a red warning message and an explicit question to warn the user about the possible loss. I think that's enough and that 'mixed' or 'soft' reset in the special case of using the reflog make no sense because, when you use it, that's because you mess something and want to reset a branch to a previous commit ditching all your current mess...

Radrik5 commented:
Personally I prefer viewing reflog commits in the main graph: in addition to date and the commit message I can view diff with the current commit and diff between two reflog commits. Having that I can easily find the right commit in reflog and either reset one file to the state of reflog commit or reset current branch using the standard reset dialog. Currently I do it by writing --rerlog in the branch name filter (it goes directly to the git log command line), the only disadvantage is that I don't see reflog names such as HEAD@{1} or master@{2}.

l-inc commented:
**pmiossec**
> I have included your fix (even if I don't understand everything...)

I'm confused. It's a simple fix from someone (me) who never wrote a single line of code in C# before. I didn't even expect it to remain the way I did it, because I thought there might have been just a property allowing to fix the persisting row selection. Or is it [mono](https://en.wikipedia.org/wiki/Mono_(software)) that caused the misunderstanding?
> I think that's enough and that 'mixed' or 'soft' reset in the special case of using the reflog make no sense

Well, as I said, I have already lost some uncommited data before. So you can safely assume that "Mixed" and "Soft" would be quite helpful for at least some use cases. In particular when you have untracked files that don't even get stashed by default. And this default is something many people would prefer to keep as is [to not lose some other data](http://blog.icefusion.co.uk/git-stash-can-delete-ignored-files-git-stash-u/).

But thank you for adding the warning anyway. That's a good start. :-)

**Radrik5**
That's a nice discovery. Thanks for sharing.

l-inc commented:
The pull request has been merged, but it's still dangerous, because the warning is not shown for some important use cases. E.g., if you unintentionally amend a commit (instead of making a new one), using the reflog window to undo the amendment will kill your changes. So I'd still fallback to the **Radrik5**'s suggestion, which allows (and luckily also defaults to) doing a mixed reset and keeping changes safe.

jbialobr commented:
I think that the existing FormResetCurrentBranch could be reused here.

l-inc commented:
Yes, that would be my preference as well.

pmiossec commented:
I'm doing it at the moment...

On 14 November 2016 at 23:27, l-inc <notifications@github.com> wrote:

> Yes, that would be my preference as well.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/3242#issuecomment-260483795>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcFpIAMJ480yvIwulTEaYdXbCbjP4tuks5q-OBBgaJpZM4JGnRI>
> .
>

kikijiki commented:
Looks like this has been merged but I can't see the reflog option in any menu, am I missing something?

pmiossec commented:
That was merged but not yet released...

On 29 May 2017 7:31 a.m., "キキジキ" <notifications@github.com> wrote:

> Looks like this has been merged but I can't see the reflog option in any
> menu, am I missing something?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/3242#issuecomment-304575729>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcFpJkq6-LWCjvRDCuhKCq0uOT7xPTXks5r-lg3gaJpZM4JGnRI>
> .
>
